clang-tools-extra/clang-tidy/readability/UseStdMinMaxCheck.cpp

```diff
@@ -0,0 +1,238 @@
+          ((CondLhs->getType()->getUnqualifiedDesugaredType() !=
+            CondRhs->getType()->getUnqualifiedDesugaredType())
+               ? "<" + GlobalImplicitCastType.getCanonicalType().getAsString() +
+                     ">("
+               : "(") +
+          CondLhsStr + ", " + CondRhsStr + ");")
+      .str();
+}
+
+UseStdMinMaxCheck::UseStdMinMaxCheck(StringRef Name, ClangTidyContext *Context)
+    : ClangTidyCheck(Name, Context),
+      IncludeInserter(Options.getLocalOrGlobal("IncludeStyle",
+                                               utils::IncludeSorter::IS_LLVM),
+                      areDiagsSelfContained()) {}
+
+void UseStdMinMaxCheck::storeOptions(ClangTidyOptions::OptionMap &Opts) {
+  Options.store(Opts, "IncludeStyle", IncludeInserter.getStyle());
+}
+
+void UseStdMinMaxCheck::registerMatchers(MatchFinder *Finder) {
+  auto AssignOperator =
+      binaryOperator(hasOperatorName("="), hasLHS(expr().bind("AssignLhs")),
+                     hasRHS(expr().bind("AssignRhs")));
+
+  Finder->addMatcher(
+      ifStmt(
+          stmt().bind("if"),
+          unless(hasElse(stmt())), // Ensure `if` has no `else`
+          hasCondition(binaryOperator(hasAnyOperatorName("<", ">", "<=", ">="),
+                                      hasLHS(expr().bind("CondLhs")),
+                                      hasRHS(expr().bind("CondRhs")))
+                           .bind("binaryOp")),
+          hasThen(
+              anyOf(stmt(AssignOperator),
+                    compoundStmt(statementCountIs(1), has(AssignOperator)))),
+          hasParent(stmt(unless(ifStmt(hasElse(
+              equalsBoundNode("if"))))))), // Ensure `if` has no `else if`
+      this);
+}
+
+void UseStdMinMaxCheck::registerPPCallbacks(const SourceManager &SM,
+                                            Preprocessor *PP,
+                                            Preprocessor *ModuleExpanderPP) {
+  IncludeInserter.registerPreprocessor(PP);
+}
+
+void UseStdMinMaxCheck::check(const MatchFinder::MatchResult &Result) {
+  const auto *If = Result.Nodes.getNodeAs<IfStmt>("if");
+  const auto &Context = *Result.Context;
+  const auto &LO = Context.getLangOpts();
+  const auto *CondLhs = Result.Nodes.getNodeAs<Expr>("CondLhs");
+  const auto *CondRhs = Result.Nodes.getNodeAs<Expr>("CondRhs");
+  const auto *AssignLhs = Result.Nodes.getNodeAs<Expr>("AssignLhs");
+  const auto *AssignRhs = Result.Nodes.getNodeAs<Expr>("AssignRhs");
+  const auto *BinaryOp = Result.Nodes.getNodeAs<BinaryOperator>("binaryOp");
+  const auto BinaryOpcode = BinaryOp->getOpcode();
+  const auto OperatorStr = BinaryOp->getOpcodeStr();
+  const SourceManager &Source = Context.getSourceManager();
+  const SourceLocation IfLocation = If->getIfLoc();
+  const SourceLocation ThenLocation = If->getEndLoc();
+  bool Found = false;
+  clang::QualType GlobalImplicitCastType;
+
+  ExprVisitor Visitor(Result.Context);
+  Visitor.visitStmt(If, Found, GlobalImplicitCastType);
```

Comment:
Do you really need to look for a `GlobalImplicitCastType` every time there is a match? 
It would be better to move this call to the `createReplacement` method and only call it if both types are not equal.

Reply:
yes great suggestion , I moved it in the createReplacement